### DISCUSSION TITLE
Grouping - Confusion over generated results on dashboard

### DISCUSSION THREAD

**Discussion by parky128** [sorry-cypress/sorry-cypress#397] [General]
I have just started using sorry-cypress and I am encountering an issue where it appears I am getting the same specs run on each node.

For example I have a couple of nodes being used for our test stage in our Jenkins Pipeline

```
stage('Cypress 1') {
        agent { label 'aws_slave_cypress' }
        steps {
            runE2eTests()
        }
    }
    stage('Cypress 2') {
        agent { label 'aws_slave_cypress' }
        steps {
            runE2eTests()
        }
    }
}
```

This in turn executes the following on each node

`sh "CYPRESS_API_URL=\"${CYPRESS_API_URL}\" npx cy2 run --record --key XXX --group 2x-chrome --parallel --ci-build-id ${env.BUILD_ID} --browser chrome --headless --config-file cypress/cypress.${E2E_ENVIRONMENT}.json --env $envCredentials || true"`

I have actually only got one spec file in the project current, just whilst I get sorry-cypress setup in our environment.

But I see in the dashboard the same spec gets run twice, one for each node, I can see different machine id's when I click into each of these runs.

Why are these not getting grouped together, what am I missing? 

![image](https://user-images.githubusercontent.com/10373419/125780208-06ba481c-ba00-49be-ac90-c8d4d336fe6c.png)

**Comment by agoldis** [comment]
Please make sure you're using the same buildId If you want two invocation to be within the same run. 
Another possible issue is missing indexes on MongoDB "runs" collection - they might disappear in some edge cases and I still didn't have chance to fix this

**Reply by parky128** [reply]
Well as per the command sample I posted in my question, I am using the same build ID value via the Jenkins `env.BUILD_ID` variable in my pipeline script - unless I am supply that incorrectly? Can't see how as the correct build number from Jenkins is making its way through to the dashboard

**Reply by agoldis** [reply]
Yeah, they have the same buildId indeed.

Can you please make sure that the indexes are defined properly on your mongo instance as appears here: https://github.com/sorry-cypress/sorry-cypress/blob/0ec6ea72759be59a98a28b1fce75ee46564e187f/packages/mongo/src/db.ts#L51

**Reply by agoldis** [reply]
I've just run a grouped tests on v1.0.3 https://sorry-cypress-demo.herokuapp.com/run/5692bc05d2ecc95e92db5a4356cb0cdb - it all works fine, so it must be something with the configuration

**Reply by parky128** [reply]
OK so I set up our sorry-cypress stack on AWS using the cloud formation template from here https://github.com/sorry-cypress/sorry-cypress/wiki/AWS-Tutorial

So could it be something up with the template, or maybe I missed a step when following the stack setup in my AWS?

**Reply by agoldis** [reply]
It usually works well, I'd suggest starting with simple use case, i.e. no grouping first, see that parallelization works, then adding grouping etc.

**Reply by parky128** [reply]
Yeah thats exactly the order I have started using this. So no grouping to begin with. I only have one test spec in my project as a proof of concept, and two nodes being run in parallel passing the same buildID from Jenkins, but for some reason the same spec is getting claimed separately by each node, its like the director service is not aware of them both. Should I be doing anything special to instruct what specs should be run on which node or should that get handled for me by this service?

**Reply by agoldis** [reply]
> Should I be doing anything special

No, it'd be seamless and the parallelization should work out of the box for those run. Were you able to verify the index on mongo?

You can create a test runs with a demo project to see if there's any difference in the setup, I could also take a look at the logs and the DB enries

```
CYPRESS_API_URL=https://sorry-cypress-demo-director.herokuapp.com/ cy2 run --record --key blahblah --parallel --ci-build-id whatever
```

**Reply by parky128** [reply]
Ok, so I have used the demo url you supplied ion your previous comment and have performed a couple of runs on our Jenkins, one with and one without the group argument. 

See runs here https://sorry-cypress-demo.herokuapp.com/o3-portero/runs

I have two nodes being used during the build job and only one spec file in total.

So I'm seeing the same behaviour with our own sorry-cypress instance. If you can look into the logs and tell me what Im doing wrong configuration wise then I will be forever in your debt!

Thanks

**Reply by agoldis** [reply]
@parky128 
Thanks for generating those runs. 

They have different git commits thus being recognized as different runs:
https://github.com/sorry-cypress/sorry-cypress/blob/0ec6ea72759be59a98a28b1fce75ee46564e187f/packages/director/src/lib/hash.ts#L10

- https://sorry-cypress-demo.herokuapp.com/run/3516029de03353214c95ab72578318fe points to sha `754c72bd401cf67f327449640a0dffee5338d02e`
- https://sorry-cypress-demo.herokuapp.com/run/5907d8ad59b91e21739fed089c11b533 points to sha `efb52dd25391294f0d0074522667759696995e89`

**Reply by parky128** [reply]
Oh right, well that's fried my brain! I'll go figure out why that's the case, but thanks for the explanation

**Reply by parky128** [reply]
Hi @agoldis 

I am just coming back to this after some vacation, and looking closer at why different commits are being generated for each run.

I am using Jenkins and have a pipeline setup with a parallel stage configured to spin up a couple of AWS EC2 instances and run the cypress test commands against. I can see in the generated Jenkins build output that for each parallel stage, its cloning my repo and creating independent merge commits which is why they are getting treated as separate runs.

Not sure how familiar you are with Jenkins pipelines, but I have this setup right now as a stage:

```
stage('E2E-PR') {
    parallel {
        stage('Cypress Tests 1') {
            agent { label 'aws_slave_cypress' }
            steps {
                runE2eTests()
            }
        }
        stage('Cypress Tests 2') {
            agent { label 'aws_slave_cypress' }
            steps {
                runE2eTests()
            }
        }
    }
}
```

I have not yet figured out how to get around this, one thing I am trying out is getting the parallel runs to use the same original generated workspace from the start of my pipeline on my master (parent) node.

Do you have any Jenkins experience and perhaps some insight into what I could\should do here?

Thanks

**Reply by parky128** [reply]
Ok, ignore my previous message above, I am now seeing the runs grouped together, no longer getting the different commit ids causing me the problem. I think it was something to do with my slave nodes getting messed up with my running all manner of experiments on them, i just nuked them and got new instances spun up on my Jenkins, and now its all behaving.
